### ISSUE TITLE
Refactor `--screenshot-high-bit-depth` to be more robust?

### ISSUE THREAD

**Issue by dokur0chan** [mpv-player/mpv#12583]
### Expected behavior of the wanted feature

Add the ability to control the pixel format for image formats that support it, such as PNGs. For example, with PNGs, `--screenshot-high-bit-depth` currently has four different behaviors depending on the VO.

Using the following will generate an RGB24 screenshot:
```
--screenshot-format=png --screenshot-high-bit-depth=no
```
Using the following will generate an RGBA32 screenshot:
```
--screenshot-format=png --screenshot-high-bit-depth=yes
```
(**Note:** I'm pretty sure this actually a `vo_gpu` bug, because the image is tagged as RGBA64, but the file size doesn't actually reflect this...)

As for `vo_gpu_next`, the first command will generate an RGBA32 screenshot and the second command will generate an RGBA64 screenshot (and unlike with `vo_gpu`, I'm pretty sure this is a _true_ RGBA64 screenshot). 3 out of 4 of these scenarios generate a screenshot with an intact alpha channel, which some users might find undesirable since typical video content is not encoded as YUVA, and the extra alpha channel contributes to the file size.
### Alternative behavior of the wanted feature

Simply add an option to throw out the alpha channel instead of letting users choose specific pixel formats (this would basically accomplish the same thing, but would allow `--screenshot-high-bit-depth` to stay as a `yes`/`no` option).

If there's a valid reason to keep the alpha channel in every scenario, even for screenshots of YUV video, then this FR is obviously pointless and can be closed.

**Comment by sfan5** [comment]
The image_writer code is indeed in a good place to throw out the alpha channel (e.g. with a new option as you mentioned) but I think it's up to the VOs to not return an useless alpha channel.

mpv does have an option similar to what you initially described (`--screenshot-avif-pixfmt`)
but I think it's bad design because it exposes way too many implementation details, interchangeable fmts like nv12 and nv21, deprecated stuff like yuvj, endianness.
Meanwhile actual use cases look more like alpha yes/no?, chroma 444/422/420?, high depth yes/no?
The only reason it still exists is that I couldn't yet be bothered to implement something better.

> Using the following will generate an RGB24 screenshot:
> 
> --screenshot-format=png --screenshot-high-bit-depth=no
> 
> Using the following will generate an RGBA32 screenshot:
> 
> --screenshot-format=png --screenshot-high-bit-depth=yes

This *should* not be happening. If the VO returned an alpha channel it should be consistently kept (or there should be none if there wasn't one).

**Comment by dokur0chan** [comment]
> Meanwhile actual use cases look more like alpha yes/no?

Yeah, on second thought, just having an option to disregard the alpha channel is probably fine.

> This should not be happening. If the VO returned an alpha channel it should be consistently kept (or there should be none if there wasn't one).

I'm not really sure what causes the discrepancy. `--screenshot-high-bit-depth` does nothing for 8-bit encodes (it saves screenshots as RGB24 regardless, as noted here by haasn https://github.com/mpv-player/mpv/issues/9505#issuecomment-1261603064). But for 10-bit encodes, `vo_gpu` saves the screenshot as `Format: rgba64`.

Random mediainfo babble of the same frame in a 10-bit encode:

### vo_gpu + screenshot-high-bit-depth=yes
```
Complete name                            : mpv-shot0001.png
Format                                   : PNG
Format/Info                              : Portable Network Graphic
File size                                : 913 KiB

Image
Format                                   : PNG
Format/Info                              : Portable Network Graphic
Compression                              : Deflate
Width                                    : 768 pixels
Height                                   : 576 pixels
Color space                              : RGBA
Bit depth                                : 16 bits
Compression mode                         : Lossless
Stream size                              : 913 KiB (100%)
```
### vo_gpu + screenshot-high-bit-depth=no
```
Complete name                            : mpv-shot0002.png
Format                                   : PNG
Format/Info                              : Portable Network Graphic
File size                                : 622 KiB

Image
Format                                   : PNG
Format/Info                              : Portable Network Graphic
Compression                              : Deflate
Width                                    : 768 pixels
Height                                   : 576 pixels
Color space                              : RGB
Bit depth                                : 8 bits
Compression mode                         : Lossless
Stream size                              : 622 KiB (100%)
```

### vo_gpu_next + screenshot-high-bit-depth=yes
```
Complete name                            : mpv-shot0003.png
Format                                   : PNG
Format/Info                              : Portable Network Graphic
File size                                : 2.34 MiB

Image
Format                                   : PNG
Format/Info                              : Portable Network Graphic
Compression                              : Deflate
Width                                    : 768 pixels
Height                                   : 576 pixels
Color space                              : RGBA
Bit depth                                : 16 bits
Compression mode                         : Lossless
Stream size                              : 2.34 MiB (100%)
```

### vo_gpu_next + screenshot-high-bit-depth=no
```
Complete name                            : mpv-shot0004.png
Format                                   : PNG
Format/Info                              : Portable Network Graphic
File size                                : 690 KiB

Image
Format                                   : PNG
Format/Info                              : Portable Network Graphic
Compression                              : Deflate
Width                                    : 768 pixels
Height                                   : 576 pixels
Color space                              : RGBA
Bit depth                                : 8 bits
Compression mode                         : Lossless
Stream size                              : 690 KiB (100%)
```

So yeah TL;DR I dunno what's up with that. It might also be worth noting that for the `d3d11` backend, `vo_gpu` selects the 8-bit backbuffer (`R8G8B8A8_UNORM`) and `vo_gpu_next` selects the 10-bit one (`R10G10B10A2_UNORM`). I dunno if that actually changes anything for GPU screenshots and could explain some of the disparities in the file sizes.

**Comment by sfan5** [comment]
My test results on a source file with yuv420p10:

vo | high-bit-depth | result
--- | ---- | ---
gpu | yes | PNG image data, 1920 x 1080, 16-bit/color RGBA, non-interlaced
gpu | no | PNG image data, 1920 x 1080, 8-bit/color RGB, non-interlaced
gpu-next | yes | PNG image data, 1920 x 1080, 16-bit/color RGBA, non-interlaced
gpu-next | no | PNG image data, 1920 x 1080, 8-bit/color RGBA, non-interlaced

I can also confirm at the gpu-next 10-bit one is much larger than the gpu one, which is indeed suspicious.

**Comment by sfan5** [comment]
Mystery solved: vo_gpu also changes its behavior based on the options set.
https://github.com/mpv-player/mpv/blob/d9b33dd79ad1f5e0ea7b4a44592e73c905f81ece/video/out/gpu/video.c#L3499-L3507
Getting rid of alpha for high-depth screenshots is not trivial here since there exists no pixel format like rgb0 with 64bpp.

As for the size difference vo_gpu still applies the dithering to screenshots, compare the result with `--dither-depth=no`.
Can probably be fixed. I'll look into this.
